_posts/2021-02-12-aprendendo-programacao-funcional-com-elixir-cap-2.md

```diff
@@ -0,0 +1,530 @@
+
+* Os valores de boolean e nil são considerados atoms (`:nil == nil`, `:true == true`, `:false == false` )
+
+* (Rachel: em elixir a gente não pode gerar tipos novos, como acontece em OO. No máximo você cria umas structs, que é um map mais avançado, que será visto mais pra frente)
+
+## Executando código e gerando um resultado (operadores)
+
+* Quando usamos o terminal (iex), digitamos e apertamos enter o computador vai processar aquele valor.
+
+* Se você digitar `42` ele vai retornar `42` da mesma forma que se você digitar `1+1` ele retorna `2`
+
+```elixir
+iex> 42
+42
+iex> 1+1
+2
+iex> (2+2) * 3 # <---- vai respeitar as regras matemáticas
+12
+iex> 2 + 2 * 3
+8
+```
+
+* Geralmente quando você tenta criar expressões com tipos não compatíveis, ele retorna um erro
+
+```bash
+iex> "olar" + 33
+** (ArithmeticError) bad argument in arithmetic expression
+    :erlang.+("olar", 33)
+```
+
+* Mas não podemos pensar que usar operadores em tipos diferentes pode sempre dar ruim. Quando juntamos  `float` com `integer`, funciona
+
+```elixir
+iex> 12 + 8.5
+20.5
+```
+
+* O exemplo acima porque um seja `integer` e o outro seja `float`, o elixir ver ambos como `number`, que é a junção de ambos os tipos
+
+### Operadores (adicionado por Rachel)
+
+* A maioria dos operadores que conhecemos de outras linguagens funciona também em elixir
+
+* `+`, `-`, `/`, `*` para fazer operações aritméticas entre números, de adição subtração, divisão e multiplicação respectivamente
+
+* `==`, `!=`, `<`, `<=`, `>`, `>=` para comparar dois valores e retornar um boolean. Respectivamente ele compara se os valores são iguais, diferentes, menor, menor igual, maior e maior igual (`:nil == nil`, por exemplo (vai retornar true, inclusive))
+
+* Elixir tem dois operadores que não são tão comuns em outras linguagens, que é o `++` e o `<>` que servem para concatenar
+
+* o `++` serve para concatenar listas
+
+```elixir
+iex> [1,2] ++ [3,4]
+[1,2,3,4]
+```
+
+* enquanto o `<>` junta strings
+
+```elixir
+iex>"Sorvete " <> "de " <> "Creme."
+"Sorvete de Creme."
+```
+
+* vale frisar aqui que o operador `+` não funciona para juntar strings, apenas o `<>`
+
+```elixir
+iex> "Sorvete " + "de " + "Creme."
+** (ArithmeticError) bad argument in arithmetic expression: "Sorvete " + "de "
+    :erlang.+("Sorvete ", "de ")
+```
+
+### Criando expressões lógicas
+
+* Elixir também tem operadores para comparar se duas ou mais expressões são verdadeiras ou falsas
+
+* `and` e `&&` retorna true se as duas expressões forem verdadeiras. Se uma for falsa, ele retorna falso `1+1 == 2 && 2 + 2 == 4` vai retornar `true`, por exemplo, enquanto `1 + 1 == 3 && 1 + 1 == 2` vai retornar `false` (o `&&` pode ser substituído por `and`)
```

Comment:
Aqui vale diferenciar o `and` e `&&`, `or` e `||`, o que vc acha? 😃 
Algo na linha:

O `and` e `or` só funcionam com booleanos e retornam sempre booleanos.
O `&&` e `||` funcionam com expressões truthy e falsy. Por exemplo, `is_integer(2) && 3`, retorna 3, `is_nil(2) && 3`, retorna false (que é o primeiro valor). O `||` vai ser o contrário, se o primeiro for true, retorna o primeiro valor.
`is_integer(2) || 3`, retorna true, `is_nil(2) || 3`, retorna 3.

Vou ver se penso em outro exemplo